### ISSUE TITLE
#[deny(unknown_lints)] doesn't apply to code from proc_macro 

### ISSUE THREAD

**Issue by peter-kehl** [rust-lang/rust#109881]
Code generated from `proc_macro` (or parts coming from `proc_macro` and incorporated (invoked by) a macro by example) doesn't get checked 
by `#[deny(unknown_lints)]` (and friends). Can it/should it; if not, could we document this?

That is even more confusing, because code coming from pure `macro_rules!`-based macros IS checked by `#[deny(unknown_lints)]`.

For an example, see https://github.com/peter-kehl/nonexisting_lint_from_proc_macro/blob/main/proc_mac/src/lib.rs for the proc macro, and https://github.com/peter-kehl/nonexisting_lint_from_proc_macro/blob/main/use_cases/src/lib.rs for its "consumer".

(Same on `1.68.2` and `1.70.0-nightly`.)

**Comment by clubby789** [comment]
This isn't proc-macro specific, the lint is just configured not to fire on code inside macros in external crates (missing `report_in_external_macro`) - if the `macro_rules` macro is defined upstream, there is also no warning.

@rustbot label +A-lints

**Comment by peter-kehl** [comment]
Thank you @clubby789. I'm updating the issue name respectively.

**Comment by clubby789** [comment]
> Is this determined only by span
Yes, the lint is silenced if `err.span.primary_spans().iter().any(|s| in_external_macro(sess, *s))`

**Comment by peter-kehl** [comment]
If a non-existing lint comes from/through [`paste`](https://crates.io/crates/paste) crate ([`dtolnay/paste`](https://github.com/dtolnay/paste/blob/master/src/lib.rs)), the lint DOES get reported. I've updated my example.

Unfortunately, I can't use `paste`, because `paste` can generate `Ident` only, but I need to generate `Path`. I'll investigate if `paste` generates anything else different to how I generate.

**Comment by peter-kehl** [comment]
Summary: The above problem was caused by (my) using `Span::call_site()` when generating `Ident`(s) for the lint name (or the lint path, if prefixed with `clippy` or `rustdoc`).

Solution: Use a `Span` object coming from the macro input. It's enough to use such a `Span` for the `Ident`(s) covering the lint name (or path). The rest of the generated code (that is: `#[allow(` and its matching `)]`) can use `Span::call_site()`.

For prefixless (standard) lints (as opposed to `clippy::` or `rustdoc::` lints) you can use `paste` crate. (That doesn't work for `clippy::` or `rustdoc::` lints, because their lint is a path rather than just one identifier). See https://github.com/peter-kehl/nonexisting_lint_from_proc_macro/blob/main/use_cases/src/lib.rs#L18.

For the proof that a generated lint name fails to be reported as an incorrect lint when using `Span::call_site()`, you could modify `1.0.12` (the current version) of `paste` crate: https://github.com/dtolnay/paste/blob/58a4a8f86aad820d58010bf2377405f2bf2f56fd/src/lib.rs#L442 > `fn pasted_to_tokens` > change `let ident = match panic::catch_unwind(|| Ident::new(&pasted, span)) {...}` to `let ident = match panic::catch_unwind(|| Ident::new(&pasted, Span::call_site())) {...}` and the above example stops failing for incorrect lint names.

**Comment by peter-kehl** [comment]
Could this be documented at `Span::call_site()`, and/or anywhere else, please? Because as-is, this is highly confusing/misleading. From documentation of `Span::call_site()` one would expect it to be interchangeable with `Span` coming from the macro input stream (except for a source code position shift).

`grep -r in_external_macro` in rust's sources shows some 163 occurrences, so could this affect other use cases?
